x-pack/libbeat/management/generate.go

```diff
@@ -142,7 +146,7 @@ func injectIndexStream(expected *proto.UnitExpectedConfig, inputType string, str
 	}
 
 	namespace := DefaultNamespaceName
-	if testNamespace := expected.GetDataStream().GetNamespace(); testNamespace != "" {
+	if testNamespace := streamExpected.GetDataStream().GetNamespace(); testNamespace != "" {
```

Comment:
I think we may still be missing cases with this parsing. I think the precedence order needs to be:

1. If no data stream information is specified, the default should be `$defaultType-generic-default` which will typically be one of `logs-generic-default` or `metrics-generic-default` depending on the beat. This is described in the data stream naming scheme [blog post](https://www.elastic.co/blog/an-introduction-to-the-elastic-data-stream-naming-scheme), where the `type` has no default, `dataset` defaults to `generic`, and `namespace` defaults to `default`.
2. Look at the input level `data_stream` field, and override the defaults above with any values specified there. Almost always the `namespace` is specified at this level, and I believe users can edit this for some integrations so we need to consider it.
3. Look at the stream level `data_stream` field, and take any values specified there as the final values.

Here's a few examples of I think this should work:
```yaml
- id: input1             # Sets data stream fields to logs-generic-default
  type: synthetics/icmp
  streams:
    - id: stream1
- id: input1             # Sets data stream fields to synthetics-icmp-default
  type: synthetics/icmp
  data_stream:
    namespace: default
  streams:
    - id: stream1
      data_stream:
        dataset: icmp
        type: synthetics
- id: input1             # Sets data stream fields to synthetics-example-namespace
  type: synthetics/icmp
  data_stream:
    namespace: namespace
    dataset: example
  streams:
    - id: stream1
      data_stream:
        type: synthetics
```

Reply:
This is what the V1 agent was doing in 8.5, which is almost what I described above except it doesn't explicitly pull the `type` field out of `input.streams.stream[n].data_stream.type` it takes it as an input parameter: https://github.com/elastic/elastic-agent/blob/0a6c4f05e33ed054ce54409a59f6d7da52e07cd2/internal/pkg/agent/transpiler/rules.go#L587-L634

Also the `inject_index` rule wasn't used in every Beat spec file in V1, notably Heartbeat doesn't use it.

Reply:
Clearly we need a spec on how to parse an agent policy because this is very inconsistent.

Reply:
Heartbeat does its own parsing here: https://github.com/elastic/beats/blob/fd82372d4b0f520bdafa9338537e48b9d7bb7967/heartbeat/monitors/factory.go#L284-L303

Reply:
> Clearly we need a spec on how to parse an agent policy because this is very inconsistent.

Agreed. I ran into a lot of issues while testing this because of how allover some of the integrations are, this whole system could be much more formalized.